### ISSUE TITLE
Cache Brush creation under Direct2D

### ISSUE THREAD

**Issue by amerkoleci** [AvaloniaUI/Avalonia#98]
At the moment, brushes are created and disposed during drawing using the C# "using" pattern, don't know how Direct2D manages creation and dispose of Direct2D resource, would be good to cache brushes per render target and dispose all of them when target or Brush gets disposed.

**Comment by grokys** [comment]
Thanks for submitting this issue, I've been meaning to log it for a long time!

Yes, this definitely needs doing. There are a few techniques we could consider so I've been putting it off until we have a decent-size app with which to test.

**Comment by amerkoleci** [comment]
Nope, I'll try to implement this and report pull request.

**Comment by grokys** [comment]
Ok, a few thoughts on this:

- Brushes don't hold a reference to their platform-specific impl
- Even if they did, it's not feasable to expect all brushes to be destroyed when they're no longer needed
- The cache policy for different brushes will be different - i.e. we could use the same brush for all solid color brushes with the color #ff808080, the impl for VisualBrush will only be removed from the cache when the Visual is updated etc etc

**Comment by grokys** [comment]
FYI @ncarrillo is currently working on this, so I'd speak to him before starting anything if you've not already started.

**Comment by ncarrillo** [comment]
I've been looking into using a `MemoizingMRUCache` from [Splat](https://github.com/paulcbetts/splat/blob/master/Splat/MemoizingMRUCache.cs) in order to cache `SolidColorBrush` and potentially `LinearGradientBrush`.

An article on `MemoizingMRUCache` is [here](http://blog.paulbetts.org/index.php/2010/07/13/reactivexaml-series-using-memoizingmrucache/).
